### PR TITLE
Add checkout method in Ensure handler

### PR DESCRIPTION
**Pull request by nicholasSUSE**
GitHub Issue - https://github.com/rancher/rancher/issues/43021
GitHub Issue - https://github.com/rancher/rancher/issues/39532
 
# Problem
The release blocker was happening because of the solution implemented for this old bug:
This is the fix implemented at the time:https://github.com/rancher/rancher/issues/39532
This is the fix implemented at the time: https://github.com/rancher/rancher/pull/42367

This was breaking follower pods from resetting to the right commit.

# Solution
This was breaking follower pods from resetting to the right commit.
We changed the hardReset approach to a checkout-to-commit approach.
This way we ensure the right behavior without causing the old bug (checkout to commit changes to the work tree but does not erase the index log)

### CODE REVIEW

File: pkg/catalogv2/git/download.go [hunk 1]
```diff
@@ -6,25 +6,24 @@ import (
 	plumbing "github.com/go-git/go-git/v5/plumbing"
 )
 
-// Ensure will check if repo is cloned, if not the method will clone and reset to the latest commit.
-// If reseting to the latest commit is not possible it will fetch and try to reset again
-func (r *Repository) Ensure(branch string) error {
-	// clone at the current HEAD pointing branch
-	// if the HEAD pointing branch is supposed to change, then it should be done at Update or Head method
+// Ensure will check if repo is cloned, if not the method will clone and checkout to the ClusterRepo.Status.Commit.
+// If checking out to the given commit is not possible it will fetch and try to checkout again
```
Comment (rohitsakala) [inline pkg/catalogv2/git/download.go:10]: it will fetch and reset

File: pkg/catalogv2/git/download.go [hunk 2]
```diff
@@ -6,25 +6,24 @@ import (
...
+func (r *Repository) Ensure(commit string) error {
+	// fresh clone or open
+	// if the HEAD pointing commit is supposed to change, then it should be done at Update or Head method
 	err := r.cloneOrOpen("")
 	if err != nil {
 		return fmt.Errorf("failed to clone or open repository: %w", err)
 	}
 
 	// Try to reset to the given branch, if success exit
```
Comment (rohitsakala) [inline pkg/catalogv2/git/download.go:19]: Try to checkout

Comment (rohitsakala) [reply pkg/catalogv2/git/download.go:19]: try to checkout at a given commit

File: pkg/catalogv2/git/git.go [hunk 3]
```diff
@@ -269,14 +269,32 @@ func (r *Repository) getCurrentCommit() (plumbing.Hash, error) {
...
+	err = wt.Checkout(&checkOpts)
+	if err != nil {
+		return fmt.Errorf("checkout failure: %w", err)
+	}
+
+	return nil
+}
+
 // fetchAndReset is a convenience method that fetches updates from the remote repository
 // for a specific branch, and then resets the current branch to a specified commit.
```
Comment (rohitsakala) [inline pkg/catalogv2/git/git.go:292]: Update the docs

File: pkg/catalogv2/git/git.go [hunk 4]
```diff
@@ -285,27 +303,32 @@ func (r *Repository) fetchAndReset(branch string) error {
 //   - Otherwise, it sets the RefSpec to fetch all branches.
 //
 // fetching the last commit of one branch is faster than fetching from all branches.
```
Comment (rohitsakala) [inline pkg/catalogv2/git/git.go:305]: update the docs

File: pkg/catalogv2/git/git.go [hunk 5]
```diff
@@ -285,27 +303,32 @@ func (r *Repository) fetchAndReset(branch string) error {
...
 	var newRefSpec string
 
-	if branch != "" {
-		newRefSpec = fmt.Sprintf("+refs/heads/%s:refs/remotes/origin/%s", branch, branch)
-	} else {
-		newRefSpec = "+refs/heads/*:refs/remotes/origin/*"
+	mode := checkReference(reference)
+	switch mode {
+	case CommitMode:
+		newRefSpec = fmt.Sprintf("+%s*:refs/remotes/origin/*", reference)
```
Comment (rohitsakala) [inline pkg/catalogv2/git/git.go:312]: Remove the star after string

File: pkg/catalogv2/git/git.go [hunk 6]
```diff
@@ -285,27 +303,32 @@ func (r *Repository) fetchAndReset(branch string) error {
...
+	return nil
 }
 
 // fetch fetches updates from the remote repository for a specific branch.
 // If the fetch operation is already up-to-date, this is not treated as an error.
 // Any other error that occurs during fetch is returned.
-func (r *Repository) fetch(branch string) error {
-	r.updateRefSpec(branch)
+func (r *Repository) fetch(reference string) error {
+	r.updateRefSpec(reference)
```
Comment (rohitsakala) [inline pkg/catalogv2/git/git.go:331]: check the error

File: pkg/catalogv2/git/git.go [hunk 7]
```diff
@@ -285,27 +303,32 @@ func (r *Repository) fetchAndReset(branch string) error {
...
 
 	if len(r.fetchOpts.RefSpecs) > 0 {
 		r.fetchOpts.RefSpecs[0] = config.RefSpec(newRefSpec)
 	} else {
 		r.fetchOpts.RefSpecs = []config.RefSpec{config.RefSpec(newRefSpec)}
 	}
+	return nil
 }
 
 // fetch fetches updates from the remote repository for a specific branch.
```
Comment (rohitsakala) [inline pkg/catalogv2/git/git.go:327]: Update the docs

File: pkg/catalogv2/git/utils.go [hunk 8]
```diff
@@ -40,6 +42,37 @@ func isLocalBranch(branch string) bool {
 	return strings.HasPrefix(branch, localReferenceBranch)
 }
 
+// isValidCommitHash checks if the provided input string is a valid git commit hash
+//   - true: if valid
+//   - false; if invalid
+func isValidCommitHash(input string) bool {
```
Comment (rohitsakala) [inline pkg/catalogv2/git/utils.go:48]: Can we use plumbing.Ishash() ?

File: pkg/controllers/dashboard/helm/repo.go
Comment (rohitsakala) [inline pkg/controllers/dashboard/helm/repo.go:1]: Can we make sure the ensure function is back to its old behaviour ? the if conditions as well ?

File: pkg/catalogv2/git/download.go (hunk 1 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/download.go:10]: Modified

File: pkg/catalogv2/git/download.go (hunk 2 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/download.go:19]: Done

File: pkg/catalogv2/git/git.go (hunk 3 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/git.go:292]: Done

File: pkg/catalogv2/git/git.go (hunk 4 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/git.go:305]: Done

File: pkg/catalogv2/git/git.go (hunk 5 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/git.go:312]: Done

File: pkg/catalogv2/git/git.go (hunk 7 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/git.go:327]: Done

File: pkg/catalogv2/git/utils.go (hunk 8 above)
Comment (nicholasSUSE) [reply pkg/catalogv2/git/utils.go:48]: Yes

**Review by rohitsakala — DISMISSED** [review]

**Comment by rohitsakala** [comment]
Test Scenarios 

Non-AirGap
- Non HA 
1. Upgrade from 2.7 to 2.8 
2. OOB Release Updates
3. Changing branches without upgrade 

- HA
1. Upgrade from 2.7 to 2.8 
2. OOB Release Updates
3. Changing branches without upgrade 

AirGap

- Non HA
1. Upgrade from 2.7 to 2.8 
2. OOB Release Updates
3. Changing branches without upgrade 

- HA
1. Upgrade from 2.7 to 2.8 
2. OOB Release Updates
3. Changing branches without upgrade

File: pkg/catalogv2/git/git.go [hunk 9]
```diff
@@ -269,54 +269,79 @@ func (r *Repository) getCurrentCommit() (plumbing.Hash, error) {
...
-		newRefSpec = fmt.Sprintf("+refs/heads/%s:refs/remotes/origin/%s", branch, branch)
-	} else {
-		newRefSpec = "+refs/heads/*:refs/remotes/origin/*"
+	mode := checkReference(reference)
+	switch mode {
+	case CommitMode:
+		newRefSpec = fmt.Sprintf("+%s:refs/remotes/origin/*", reference)
+	case BranchMode:
+		newRefSpec = fmt.Sprintf("+refs/heads/%s:refs/remotes/origin/%s", reference, reference)
+	case "":
```
Comment (diogoasouza) [inline pkg/catalogv2/git/git.go:314]: When will this happen? The checkReference function does not return an empty string

File: pkg/controllers/dashboard/helm/repo.go [hunk 10]
```diff
@@ -134,7 +134,7 @@ func toOwnerObject(namespace string, owner metav1.OwnerReference) runtime.Object
 // checks are made to determine whether the repo exists and is ready.
 func (r *repoHandler) ensure(repoSpec *catalog.RepoSpec, status catalog.RepoStatus, metadata *metav1.ObjectMeta) (catalog.RepoStatus, error) {
 	// related ClusterRepo Status is not updated by download handler yet
-	if status.Branch == "" || status.Branch != repoSpec.GitBranch {
+	if status.Commit == "" {
```
Comment (diogoasouza) [inline pkg/controllers/dashboard/helm/repo.go:137]: Why did you remove the check of status != repoSpec ?

File: pkg/catalogv2/git/git.go [hunk 11]
```diff
@@ -269,44 +269,73 @@ func (r *Repository) getCurrentCommit() (plumbing.Hash, error) {
...
+//   - If nothing is provided or a invalid commit Hash it will throw an error.
+func (r *Repository) updateRefSpec(reference string) error {
 	var newRefSpec string
 
-	if branch != "" {
-		newRefSpec = fmt.Sprintf("+refs/heads/%s:refs/remotes/origin/%s", branch, branch)
-	} else {
-		newRefSpec = "+refs/heads/*:refs/remotes/origin/*"
+	mode := checkReference(reference)
+	switch mode {
```
Comment (rohitsakala) [inline pkg/catalogv2/git/git.go:318]: Remove switch

File: pkg/controllers/dashboard/helm/repo.go (hunk 10 above)
Comment (rohitsakala) [reply pkg/controllers/dashboard/helm/repo.go:137]: To replicate the old behaviour. In release/v2.7 only status.commit was being checked and status.branch != spec.branch was not doing anything.

**Review by rohitsakala — APPROVED** [review]

File: pkg/catalogv2/git/git.go [hunk 12]
```diff
@@ -269,30 +269,51 @@ func (r *Repository) getCurrentCommit() (plumbing.Hash, error) {
...
 
 // updateRefSpec updates the reference specification (RefSpec) in the fetch options
 // of the repository operation.
 //   - If a branch name is provided, it sets the RefSpec to fetch that specific branch.
-//   - Otherwise, it sets the RefSpec to fetch all branches.
-//
-// fetching the last commit of one branch is faster than fetching from all branches.
-func (r *Repository) updateRefSpec(branch string) {
-	var newRefSpec string
+//   - If a commit is provided, it sets the RefSpec to fetch that specific commit.
```
Comment (rohitsakala) [inline pkg/catalogv2/git/git.go:312]: Update Docs

File: pkg/catalogv2/git/utils.go [hunk 13]
```diff
@@ -17,6 +17,8 @@ const (
 	staticDir            = "/var/lib/rancher-data/local-catalogs/v2"
 	localDir             = "../rancher-data/local-catalogs/v2" // identical to helm.InternalCatalog
 	localReferenceBranch = "refs/heads/"
+	CommitMode           = "commit"
```
Comment (rohitsakala) [inline pkg/catalogv2/git/utils.go:20]: Remove these

File: pkg/catalogv2/git/git.go (hunk 9 above)
Comment (rohitsakala) [reply pkg/catalogv2/git/git.go]: Removed
